Tools/gsce

```diff
@@ -0,0 +1,138 @@
+           '    '+ proj_dir + '/graphene/Pal/src/host/Linux-SGX/signer/pal-sgx-sign -libpal ' + proj_dir + 
+           '/graphene/Pal/src/host/Linux-SGX/../../../../Runtime/libpal-Linux-SGX.so -key ' + proj_dir + 
+           '/graphene/Pal/src/host/Linux-SGX/signer/enclave-key.pem -output ' + app_name + '.manifest.sgx ' + 
+           '-manifest ' + app_name + '.manifest \n')
+  # Remove signing key
+  df.write('# Removing key after signing \n')
+  # TODO
+  
+  # Overwrite Entry Point
+  df.write('ENTRYPOINT  ["/bin/bash", "/gbin/app_exec"] \n') 
+  df.close()
+
+def make_exec(path) :
+  mode = os.stat(path).st_mode
+  mode |= (mode & 0o444) >> 2    # copy R bits to X
+  os.chmod(path, mode)
+
+def gen_app_executor(app_name, bin_cmd, proj_dir) :
+  if not os.path.exists(proj_dir + "/graphene/Tools/build/tools") :
+    os.makedirs(proj_dir + "/graphene/Tools/build/tools")
+
+  e_path = proj_dir + "/graphene/Tools/build/tools/app_exec"
+  print "e_path: " + e_path
+  ef = open(e_path, "w")
+  make_exec(e_path)
+  ef.write('#!/bin/bash \n \n')
+  ef.write('cd ' + proj_dir + '/graphene/LibOS/shim/test/apps/' + app_name +'\n')
+  ef.write('# Generate EINITOKEN \n')
+  ef.write(proj_dir + '/graphene/Pal/src/host/Linux-SGX/signer/pal-sgx-get-token -output ' 
+           + app_name + '.token -sig ' + app_name + '.sig \n')
+  ef.write('# Run the application \n')
+  ef.write('SGX=1 ./' + app_name + '.manifest.sgx ' + bin_cmd + '\n')
+  
+  ef.close()
+  
+if __name__ == "__main__":
+  if len(sys.argv) < 3:
+    print "Usage: gsce run [Image name] "
+    exit()
+  #image_name = sys.argv[2]
+  image_name = sys.argv[-1]
+  image_match = re.match(r'([^:]*)(:*)(.*)', image_name)
+  if image_match :
+    app_name = image_match.group(1)
+  print "app_name: " + app_name
+  inspect_cmd = "sudo docker inspect --format '{{.Config.Cmd}}' " + image_name
+  res = subprocess.check_output(inspect_cmd, shell=True).strip()
+  print res
+  match = re.match(r'\[([^\s]*)\s*(.*)\]', res)
+  bin_name = match.group(1)
+  if match.group(2) :
+    bin_cmd = match.group(2)
+  else :
+    bin_cmd = ""
+  print "bin_name: " + bin_name + " bin_cmd: " + bin_cmd
+  # Store the rest arguments as docker run arguments
+  docker_str = ""
+  for i in range (2, len(sys.argv)-1) :
+    docker_str += " " + sys.argv[i]
+  
+  # print image_cmd
+  proj_dir = os.path.abspath(os.getcwd() + "/../../")
+
+  # STEP 1: Generating Dockerfile
+  gen_dockerfile(image_name, app_name, bin_name, proj_dir)
+  
+  # STEP 2: Generating entry point execute script
+  gen_app_executor(app_name, bin_cmd, proj_dir)
+
+  # STEP 3: Building new docker image with generated Dockerfile
+  os.chdir('..')
+  os.system("sudo docker build -f Tools/build/Dockerfile." + app_name + " -t gsc_" + app_name + " .\n") 
+
+  # STEP 4: Run GSC with the target app
+  os.system("sudo docker run -i -t" + docker_str +" --device=/dev/gsgx --device=/dev/isgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket gsc_"+app_name+"\n")
```

Comment:
I'm not sure whether this is secure. I guess we'd need to look at the isgx and gsgx drivers a bit more and check if it's a good idea to expose them to a potentially malicious code from the inside of a container (or just add a warning to the end-user that he should never run untrusted payloads, even inside a container).

Reply:
Yes. This adds more resources for the containerized application to access than it original has. But it's more like system admin need to be aware of this?

Reply:
I'm not concerned about OS resources, but about security. I wouldn't expect the OS admin to read all the setup code and verify whether it's secure or not, it's IMO our responsibility ;)